README.md

```diff
@@ -1,2 +1,159 @@
 # linux-prewarm-process-rfc
 WIP - For now it's just an idea of how to launch processes ahead of time for bash scripts
+
+# context / problem
+shell scripts (in any flavour/variant) are incredibly composable due to unix pipelines. 
+This led many people to develop a plethora CLI utilies for all sorts of problems out there
+(curl, grep, sed, cat - and the list goes on and on).
+
+The problem is that there is a long list of housekeeping actions that the operating system
+must perform before creating processes such as:
+
+- dynamic linker kicks in and load dependent libraries
+- allocated physical space to accomodate virtual addresses
+- read executable from disk
+- allocate pid
+- put process on OS scheduler (usually CFS)
+- just to name a few.... there is actually much more than this.
+
+Also, it's worth mentioning that while killing processes a lot of the aforementioned
+tasks must be undone by the OS which also takes time.
+
+As a result, even small shell script files tend to be noticible slower than equivalent
+programs written in other programming languages. 
+
+For instance, imagine that you are writing a AWS lambda runtime in Bash which doesn't do
+anything else other than echoing back whatever was sent through the input payload:
+
+```bash
+#!/bin/sh
+
+set -euo pipefail
+
+# function handler
+function handler () {
+  EVENT_DATA=$1
+  echo "$EVENT_DATA" 1>&2;
+  RESPONSE="Echoing request: '$EVENT_DATA'"
+
+  echo $RESPONSE
+}
+
+# Processing
+while true
+do
+  HEADERS="$(mktemp)"
+  # Get an event. The HTTP request will block until one is received
+  EVENT_DATA=$(curl -sS -LD "$HEADERS" -X GET "http://${AWS_LAMBDA_RUNTIME_API}/2018-06-01/runtime/invocation/next")
+
+  # Extract request ID by scraping response headers received above
+  REQUEST_ID=$(grep -Fi Lambda-Runtime-Aws-Request-Id "$HEADERS" | tr -d '[:space:]' | cut -d: -f2)
+
+  # Run the handler function from the script
+  RESPONSE=$($(echo "$_HANDLER" | cut -d. -f2) "$EVENT_DATA")
+
+  # Send the response
+  curl -X POST "http://${AWS_LAMBDA_RUNTIME_API}/2018-06-01/runtime/invocation/$REQUEST_ID/response"  -d "$RESPONSE"
+done
+```
+
+PS.: if we ignore shell built-ins such as `set` and `echo`, there are 8 processes being launched in which all 
```

Comment:
I'm imagining having a setup where we have just a directory of shell scripts, and then we programmatically derive metadata about the run for each of:

- with shopt
- without shopt

With that automation, we can easily add any shell script, and probably save data in a structured way to make plots that would be convincing (or not).

Reply:
+1 on having this test folder with those scripts for benchmark purposes.